contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,439 @@
+ways like Kubernetes controllers, are responsible for check pointing, 
+persisting, and recovering their own state. This problem is so common that 
+Mesos provides a ["State Abstraction"](https://github.com/apache/mesos/blob/master/include/mesos/state/state.hpp) 
+that allows frameworks to persist their state in either ZooKeeper or the 
+Mesos Replicate Log (A Multi-Paxos based state machine used by the Mesos 
+Masters). This State Abstraction is a mutable, durable dictionary where keys 
+and values are opaque strings. As controllers only need the capability to 
+persist an immutable point in time snapshot of target Object states to 
+implement a revision history, we propose to use the ControllerRevision object 
+for this purpose.
+
+``` golang
+// ControllerRevision implements an immutable snapshot of state data. Clients 
+// are responsible for serializing and deserializing the objects that contain 
+// their internal state. 
+// Once a ControllerRevision has been successfully created, it can not be updated. 
+// The API Server will fail validation of all requests that attempt to mutate 
+// the Data field. ControllerRevisions may, however, be deleted.
+type ControllerRevision struct {
+    metav1.TypeMeta
+    // +optional
+    metav1.ObjectMeta
+    // Data contains the serialized state.
+    Data runtime.RawExtension
+    // Revision indicates the revision of the state represented by Data.
+    Revision int64
+}
+```
+
+## API Server
+The API Server must support the creation and deletion of ControllerRevision 
+objects. As we have no mechanism for declarative immutability, the API server 
+must fail any update request that updates the `.Data` field of a 
+ControllerRevision Object.
+
+## Controllers 
+This section is presented as a generalization of how an arbitrary controller 
+can use ControllerRevision to persist a history of revisions to its 
+specification type Objects. The technique is applicable, without loss of 
+generality, to the existing Kuberentes controllers that have Pod as a generated 
+type.
+
+When a controller detects a revision to the target Object state of a 
+specification type Object it will do the following.
+
+1. The controller will [create a snapshot](#version-snapshot-creation) of the 
+current target Object state.
+1. The controller will [reconstruct the history](#history-reconstruction) of 
+revisions to the Object's target Object state.
+1. The controller will test the current target Object state for 
+[equivalence](#version-equivalence) with all other versions in the Object's 
+revision history.
+    - If the current version is semantically equivalent to its immediate 
+    predecessor no update to the Object's target state has been performed.
+    - If the current version is equivalent to a version prior to its immediate 
+    predecessor, this indicates a rollback.
+    - If the current version is not equivalent to any prior version, this 
+    indicates an update or a roll forward.
```

Comment:
Is this (determination of a rollback or an update) just informational, or is the controller expected to treat the two cases differently?

Also, detecting rollback is limited to the controller's `RevisionHistoryLimit`, correct?